src/clients/rust/src/lib.rs

```diff
@@ -0,0 +1,844 @@
+    #[error("pending transfer already posted")]
+    PendingTransferAlreadyPosted,
+    #[error("pending transfer already voided")]
+    PendingTransferAlreadyVoided,
+    #[error("pending transfer expired")]
+    PendingTransferExpired,
+    #[error("imported event timestamp must not regress")]
+    ImportedEventTimestampMustNotRegress,
+    #[error("imported event timestamp must postdate debit account")]
+    ImportedEventTimestampMustPostdateDebitAccount,
+    #[error("imported event timestamp must postdate credit account")]
+    ImportedEventTimestampMustPostdateCreditAccount,
+    #[error("imported event timeout must be zero")]
+    ImportedEventTimeoutMustBeZero,
+    #[error("debit account already closed")]
+    DebitAccountAlreadyClosed,
+    #[error("credit account alreday closed")]
+    CreditAccountAlreadyClosed,
+    #[error("overflows debits_pending")]
+    OverflowsDebitsPending,
+    #[error("overflows credits_pending")]
+    OverflowsCreditsPending,
+    #[error("overflows debits_posted")]
+    OverflowsDebitsPosted,
+    #[error("overflows credits_posted")]
+    OverflowsCreditsPosted,
+    #[error("overflows debits")]
+    OverflowsDebits,
+    #[error("overflows credits")]
+    OverflowsCredits,
+    #[error("overflows timeout")]
+    OverflowsTimeout,
+    #[error("exceeds credits")]
+    ExceedsCredits,
+    #[error("exceeds debits")]
+    ExceedsDebits,
+    #[error("unknown {0}")]
+    Unknown(u32),
+}
+
+#[derive(thiserror::Error, Debug, Copy, Clone)]
+#[non_exhaustive]
+pub enum Status {
+    #[error("unexpected")]
+    Unexpected,
+    #[error("out of memory")]
+    OutOfMemory,
+    #[error("address invalid")]
+    AddressInvalid,
+    #[error("address limit exceeded")]
+    AddressLimitExceeded,
+    #[error("system resources")]
+    SystemResources,
+    #[error("network subsystem")]
+    NetworkSubsystem,
+    #[error("unknown {0}")]
+    Unknown(u32),
+}
+
+#[derive(thiserror::Error, Debug, Copy, Clone)]
+#[non_exhaustive]
+pub enum PacketStatus {
+    #[error("too much data")]
+    TooMuchData,
+    #[error("client evicted")]
+    ClientEvicted,
+    #[error("client release too low")]
+    ClientReleaseTooLow,
+    #[error("client release too high")]
+    ClientReleaseTooHigh,
+    #[error("client shutdown")]
+    ClientShutdown,
+    #[error("invalid operation")]
+    InvalidOperation,
+    #[error("invalid data size")]
+    InvalidDataSize,
+    #[error("unknown {0}")]
+    Unknown(u32),
+}
+
+#[derive(thiserror::Error, Debug, Copy, Clone)]
+#[error("not found")]
+pub struct NotFound;
+
+#[derive(Debug, Copy, Clone)]
+pub struct Reserved<const N: usize>([u8; N]);
+
+impl<const N: usize> Default for Reserved<N> {
+    fn default() -> Reserved<N> {
+        Reserved([0; N])
+    }
+}
+
+fn create_packet<RustEvent, CEvent>(
+    op: u32, // TB_OPERATION
+    events: &[RustEvent],
+) -> (Box<tbc::tb_packet_t>, Receiver<CompletionMessage<CEvent>>)
+where
+    RustEvent: Copy,
+    CEvent: From<RustEvent> + 'static,
```

Comment:
I thinks for rust specifically, we should avoid any conversions here, and make sure that this crate's `pub` types have ABI compatible with Zig. 

It is fine to expose ABI over crate boundary, as our ABI is stable.

Reply:
I pushed a commit that makes these types repr(C), with the same representation as the Zig types. As of now they are still two distinct types, one emitted by the code generator, and one written by hand. The handwritten ones are distinguished by some extra rustisms: a `Reserved` helper type and `bitflags!` fields.

So the conversions are still _there_ but they are transmutes now.

It doesn't seem fully satisfying, and I'm not sure if I've accomplished what you are hoping for. In particular there are still `From` impls, and those huge generic helper functions haven't changed at all. Can you give it another look?

(Still working on the error types)

Reply:
I had two goals here:

- Expose the "standard" TigerBeetle ABI in the public API, to not unnecessary hide a detail with is in fact public ABI
- Remove two copies of the types and associated generic infrastructure. 

The first one is more important as that concerns the public interface, the second one is more of an implementation detail, so I am happy that we've got the first one. 

But I am wondering if hitting the second target as well would be easy now? Can we instruct the generator to not generates the types, and instead use hand-written ones somehow?

Reply:
Ah, great suggestions. I have changed the code to just use the hand-written Rust definitions for Account, Transfer, *Filter, AccountBalance. It lead to some good simplifications. I did though leave the codegen for the C types so I could make some assertions about the correctness of the Rust types. See what you think.

Reply:
I've committed tb_client.rs.